block/produce.go

```diff
@@ -14,72 +14,89 @@ import (
+			t.Reset(m.conf.EmptyBlocksMaxTime)
+		}
+		defer t.Stop()
 	}
 
-	// Allow the initial block to be empty
-	produceEmptyBlock := true
 	for {
 		select {
-		// Context canceled
-		case <-ctx.Done():
+		case <-ctx.Done(): // Context canceled
 			return
-		// If we got a request for an empty block produce it and don't wait for the ticker
-		case <-m.produceEmptyBlockCh:
+		case <-m.produceEmptyBlockCh: // If we got a request for an empty block produce it and don't wait for the ticker
 			produceEmptyBlock = true
-		// Empty blocks timeout
-		case <-tickerEmptyBlocksMaxTimeCh:
-			m.logger.Debug(fmt.Sprintf("No transactions for %.2f seconds, producing empty block", m.conf.EmptyBlocksMaxTime.Seconds()))
+		case <-emptyBlocksTimer: // Empty blocks timeout
 			produceEmptyBlock = true
+			m.logger.Debug(fmt.Sprintf("no transactions, producing empty block: elapsed: %.2f", m.conf.EmptyBlocksMaxTime.Seconds()))
 		// Produce block
 		case <-ticker.C:
-			err := m.produceBlock(ctx, produceEmptyBlock)
-			if errors.Is(err, types.ErrSkippedEmptyBlock) {
+			err := m.produceAndGossipBlock(ctx, produceEmptyBlock)
+			if errors.Is(err, context.Canceled) {
+				m.logger.Error("produce and gossip: context canceled", "error", err)
+				return
+			}
+			if errors.Is(err, ErrRecoverable) {
+				m.logger.Info("produce and gossip: recoverable", "error", err)
 				continue
 			}
+			if errors.Is(err, ErrNonRecoverable) {
+				m.logger.Error("produce and gossip: non-recoverable", "error", err) // TODO: flush? or don't log at all?
+				panic(fmt.Errorf("produce and gossip block: %w", err))
+			}
 			if err != nil {
-				m.logger.Error("while producing block", "error", err)
-				m.shouldProduceBlocksCh <- false
+				m.logger.Error("produce and gossip: uncategorized", "error", err)
 				continue
 			}
-			// If empty blocks enabled, after block produced, reset the timeout timer
-			if tickerEmptyBlocksMaxTime != nil {
-				produceEmptyBlock = false
-				tickerEmptyBlocksMaxTime.Reset(m.conf.EmptyBlocksMaxTime)
-			}
-
-		// Node's health check channel
+			resetEmptyBlocksTimer()
 		case shouldProduceBlocks := <-m.shouldProduceBlocksCh:
 			for !shouldProduceBlocks {
-				m.logger.Info("Stopped block production")
+				m.logger.Info("block production paused - awaiting positive continuation signal")
 				shouldProduceBlocks = <-m.shouldProduceBlocksCh
 			}
-			m.logger.Info("Resumed Block production")
+			m.logger.Info("resumed block resumed")
 		}
 	}
 }
 
-func (m *Manager) produceBlock(ctx context.Context, allowEmpty bool) error {
+func (m *Manager) produceAndGossipBlock(ctx context.Context, allowEmpty bool) error {
+	block, commit, err := m.produceBlock(ctx, allowEmpty)
+	if err != nil {
+		return fmt.Errorf("produce block: %w", err)
+	}
+
+	if err := m.gossipBlock(ctx, *block, *commit); err != nil {
+		return fmt.Errorf("gossip block: %w: %w", err, ErrNonRecoverable) // from inspecting the gossip impl, it's clear there is no point retrying
+	}
```

Comment:
@omritoptix from looking into the gossip, it doesn't seem like it's ever worth retrying this call

Reply:
can't the p2p publish fail?

Reply:
ok, I've adjusted it now 👍